modules/mastodon/index.js

```diff
@@ -0,0 +1,288 @@
+      }
+    }
+  );
+}
+
+/*
+ * fetchResource only attempts to fetch the URL given as a particular content
+ * type. This is less than what a full ActivityPub-implementing client needs to
+ * handle; following Link headers and <link> HTML tags of the correct content
+ * types would also be required.
+ */
+function fetchResource(url, fn) {
+  robotRequest({
+    url: url,
+    headers: { 'Accept': 'application/ld+json; profile="https://www.w3.org/ns/activitystreams", application/activity+json' },
+  }, (error, response, body) => {
+    if (error) {
+      fn(error);
+    } else if (response.statusCode != 200) {
+      fn(new Error('response status code is ' + response.statusCode));
+    } else if (!['application/ld+json', 'application/activity+json'].includes(
+      contentType.parse(response.headers['content-type']).type
+    )) {
+      fn(new Error('resource is not an ActivityPub record'));
+    } else {
+      try {
+        var record = JSON.parse(body);
+        if (!['Note', 'Person'].includes(record.type)) {
+          fn(new Error('record is not a Note or Person'));
+        } else {
+          fn(null, record);
+        }
+      } catch (e) {
+        fn(e);
+      }
+    }
+  });
+}
+
+function htmlToText(s) {
+  return htmlToTextMod.fromString(s, {
+    wordwrap: null,
+    ignoreHref: true,
+  });
+}
+
+function formatRecord(record, fn) {
+  if (record.type == 'Note') {
+    fetchResource(record.attributedTo, (error, userRecord) => {
+      var str = '';
+      if (error) {
+        str += '(error getting user): ';
+      } else {
+        str += sprintf('%s (@%s): ', userRecord.name, userRecord.preferredUsername);
+      }
+      if (record.sensitive) {
+        str += sprintf('[content warning] %s', record.summary);
+      } else {
+        str += htmlToText(record.content);
+      }
+      fn(str);
+    });
+  } else if (record.type == 'Person') {
+    fn(sprintf('%s (@%s): %s',
+      record.name,
+      record.preferredUsername,
+      htmlToText(record.summary),
+    ));
+  }
+}
+
+module.exports.commands = ['masto'];
+
+module.exports.init = function(b) {
+  bot = b;
+  log = bot.log;
+
+  bot.readDataFile('config.json', function(err, data) {
+    if (err) {
+      log.info('initializing mastodon module data');
+      moduleConfig = defaultConfig();
+    } else {
+      try {
+        moduleConfig = JSON.parse(data);
+      } catch(ex) {
+        moduleConfig = defaultConfig();
+      }
+    }
+    writeConfig();
+  });
+
+  var userAgent = moduleConfig && moduleConfig.userAgent || 'EuIrcBot';
+  try {
+    var ver = require('../../package.json').version;
+    userAgent += '/' + ver;
+  } catch (ex) {}
+
+  var userAgentUrl = (moduleConfig && moduleConfig.userAgentUrl
+    || 'https://github.com/euank/EuIrcBot/blob/{commitish}/modules/mastodon/info.md');
+  if (userAgentUrl.includes('{commitish}')) {
+    var commitish = 'master';
+    try {
+      commitish = execSync('git rev-parse HEAD', {'encoding': 'utf8'}).trim();
```

Comment:
Want to do `rev-parse --short HEAD` to make the user-agent not *quite* as huge?

I don't have a better suggestion, but I would like to register mild dissatisfaction with this line of code.